### PR TITLE
Fix the taints and labels

### PR DESCRIPTION
**Pull request by petrkotas**
### What this PR does
**Before this PR:**

The Taints and Labels were not properly modeled in the API.
Introducing typy mismatch

**After this PR:**
Changes the structure model to make Taints the proper record with key, value and effect
and make label to be a map

Jira: 
Link to demo recording: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Deployment: The deployment process was considered and addressed if required
- [x] Testing: New code requires new unit tests.
- [x] Documentation: Is the documentation updated? Either in the doc located in focus area, in the README or in the code itself.
- [x] Customers: Is this change affecting customers? Is the release plan considered?

### CODE REVIEW

File: api/redhatopenshift/HcpCluster/hcpCluster-models.tsp [hunk 1]
```diff
@@ -474,6 +478,18 @@ model NodePoolSpec {
   tuningConfigs?: string[];
 }
 
+/** Taint is controlling the node taint and its effects */
+model Taint {
```
Comment (miguelsorianod) [inline api/redhatopenshift/HcpCluster/hcpCluster-models.tsp:482]: The "key" attribute and the "effect" attribute should be required and not empty.

"value" is also required but it is not clear to me right now whether it can be empty or not at CS level. I will check this in CS doing some tests. For this specific attribute, if we need to change it we can do it in a follow-up MR

File: api/redhatopenshift/HcpCluster/hcpCluster-models.tsp [hunk 2]
```diff
@@ -474,6 +478,18 @@ model NodePoolSpec {
...
+  /** The key of the taint */
+  key: string;
+
+  /** The value of the taint */
+  value?: string;
+
+  /** The effect of the taint */
+  effect: string;
```
Comment (miguelsorianod) [inline api/redhatopenshift/HcpCluster/hcpCluster-models.tsp:490]: The only allowed "effect" values are: NoSchedule, PreferNoSchedule and NoExecute

We should at least document this

File: api/redhatopenshift/HcpCluster/hcpCluster-models.tsp [hunk 3]
```diff
@@ -474,6 +478,18 @@ model NodePoolSpec {
   tuningConfigs?: string[];
 }
 
+/** Taint is controlling the node taint and its effects */
+model Taint {
+  /** The key of the taint */
+  key: string;
```
Comment (miguelsorianod) [inline api/redhatopenshift/HcpCluster/hcpCluster-models.tsp:484]: There are restrictions about the accepted character sets and lengths for "key" and "value". We should document them

File: api/redhatopenshift/HcpCluster/hcpCluster-models.tsp [hunk 4]
```diff
@@ -449,13 +449,17 @@ model NodePoolSpec {
   /** Representation of a autoscaling in a node pool. */
   autoScaling?: NodePoolAutoScaling;
 
+  // This warning means to ensure good customer experience, to not create
+  // poorly defined types. However here the type literarly is the map<string, value>
+  #suppress "@azure-tools/typespec-azure-resource-manager/arm-no-record" ""
   /** Labels for the nodes */
```
Comment (miguelsorianod) [inline api/redhatopenshift/HcpCluster/hcpCluster-models.tsp:455]: We can probably improve this description a bit

I would change it to:
"K8s labels to propagate to the NodePool Nodes"

being explicit about K8s. The reason for that is to clarify that:
* It refers to K8s labels and not Azure tags
* It refers to the K8s labels associated to the Nodes of the NodePool and not of the NodePool itself
* They are propagated to the Nodes of the NodePool

Comment (petrkotas) [reply api/redhatopenshift/HcpCluster/hcpCluster-models.tsp:455]: Thank you, this makes it clear.

File: api/redhatopenshift/HcpCluster/hcpCluster-models.tsp (hunk 2 above)
Comment (petrkotas) [reply api/redhatopenshift/HcpCluster/hcpCluster-models.tsp]: I made it enum like union, so it is really clear for the user what can be assigned here.

File: api/redhatopenshift/HcpCluster/hcpCluster-models.tsp (hunk 3 above)
Comment (petrkotas) [reply api/redhatopenshift/HcpCluster/hcpCluster-models.tsp]: I created a custom `label` scalar with imposing a limits for the string to comply with k8s rules.

File: api/redhatopenshift/HcpCluster/hcpCluster-models.tsp (hunk 1 above)
Comment (petrkotas) [reply api/redhatopenshift/HcpCluster/hcpCluster-models.tsp:497]: I made only value optional, I will make the min length 3, as this is the smallest that makes sense for keys. This will force user to put value in the key.

Comment (miguelsorianod) [reply api/redhatopenshift/HcpCluster/hcpCluster-models.tsp:497]: I am trying to understand this. Why is the minimum length of a label key 3? Reading the k8s doc it does not seem to mention it: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set

File: api/redhatopenshift/HcpCluster/hcpCluster-models.tsp [hunk 5]
```diff
@@ -428,6 +428,21 @@ model NodePoolProperties {
   spec: NodePoolSpec;
 }
 
+/** label is the k8s label type used for keys, taints and values on the nodepool nodes */
+@maxLength(63)
+@pattern("^[A-Za-z0-9/-._~%!$&'()*+,;=:]+$")
+scalar label extends string;
```
Comment (miguelsorianod) [inline api/redhatopenshift/HcpCluster/hcpCluster-models.tsp:434]: I think the naming this type _label_ might be a bit confusing. It is not a label but a _label value_.
A label is a label_key/label_value pair . This is represented by Record<label_value>

File: api/redhatopenshift/HcpCluster/hcpCluster-models.tsp [hunk 6]
```diff
@@ -474,6 +493,19 @@ model NodePoolSpec {
...
 }
 
+/** Taint is controlling the node taint and its effects */
+model Taint {
+  /** The key of the taint */
+  @minLength(3)
+  key: label;
+
+  /** The value of the taint */
+  value?: label;
```
Comment (miguelsorianod) [inline api/redhatopenshift/HcpCluster/hcpCluster-models.tsp:503]: I think saying that a taint value is a "label" and a taint key is a "label" can also be confusing.

K8s Taint key and values are a different concept than K8s Label key and values

File: api/redhatopenshift/HcpCluster/hcpCluster-models.tsp (hunk 5 above)
Comment (miguelsorianod) [reply api/redhatopenshift/HcpCluster/hcpCluster-models.tsp:434]: They also have different restrictions. For example a taint key has up to 253 characters whereas a taint value has up to 63 characters

File: api/redhatopenshift/HcpCluster/hcpCluster-models.tsp (hunk 6 above)
Comment (miguelsorianod) [reply api/redhatopenshift/HcpCluster/hcpCluster-models.tsp:503]: However I see that in CS we consider them as the same, including their validation

**Comment by miguelsorianod** [comment]
I left comments around the using the name "label" in several parts. It is not accurate with how it is in K8s but in CS we don't honor that either so I guess both following both approaches is good, as long as we perform the validation that CS performs for each of them.

File: api/redhatopenshift/HcpCluster/hcpCluster-models.tsp (hunk 6 above)
Comment (petrkotas) [reply api/redhatopenshift/HcpCluster/hcpCluster-models.tsp]: I have done the changes to cover this.

File: api/redhatopenshift/HcpCluster/hcpCluster-models.tsp (hunk 5 above)
Comment (petrkotas) [reply api/redhatopenshift/HcpCluster/hcpCluster-models.tsp]: updated to reflect this

**Comment by petrkotas** [comment]
I worked on the Regular expression for validation of labels and keys. Turns out, that it would be long and hard to follow.
`^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*\/([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]$` and I am not 100% it covers all cases. WE agreed with Miguel, we leave the validation for cluster service.

**Review by mbarnes — CHANGES_REQUESTED** [review]
Can I ask you to rerun the code generation as part of this PR?

**Comment by petrkotas** [comment]
When I run the existing config I get this error:

```
fatal   | Error: --module and --module-version must both or neither be set
fatal   | Process() cancelled due to failure 
error   |   Error: Plugin go-m4-to-gocodemodel reported failure.
error   | Autorest completed with an error. If you think the error message is unclear, or is a bug, please declare an issues at https://github.com/Azure/autorest/issues with the error message you are seeing.
```

When I update according to advice in the error:
```
fatal   | Error: --module and --containing-module are mutually exclusive
fatal   | Process() cancelled due to failure 
error   |   Error: Plugin go-transform-m4 reported failure.
```

It works when I onlly set the module, but the caveat is it creates a module.  And completely changes the behavior. 
@mbarnes please see the code I have pushed as a separate commit. I am still figuring out how to change. 
I am thinking pinning to previous version of autorest.

**Comment by petrkotas** [comment]
@mbarnes I have figured it out. Fun part is that the autorest does not propagate to package.json. The way to force version is in the parameter. I love how consistent this is.

**Comment by mbarnes** [comment]
> @mbarnes I have figured it out. Fun part is that the autorest does not propagate to package.json. The way to force version is in the parameter. I love how consistent this is.

Thanks for that.  I was curious to see if I'd need to deal with generated map types in validation code.

File: api/Makefile [hunk 7]
```diff
@@ -4,7 +4,7 @@ SHELL = /bin/bash
 generate: 
 	tsp format --check "**/*.tsp"
 	tsp compile redhatopenshift/HcpCluster --warn-as-error
-	autorest autorest-config.yaml 
+	autorest --verbose --use=@autorest/go@4.0.0-preview.63 autorest-config.yaml
```
Comment (mjlshen) [inline api/Makefile:7]: Can we move this into autorest-config.yaml? E.g.

```yaml
go:
  use: autorest/go@4.0.0-preview.63
```

File: api/redhatopenshift/HcpCluster/hcpCluster-models.tsp [hunk 8]
```diff
@@ -440,6 +440,40 @@ model NodePoolProperties {
...
+@minLength(1)
+@maxLength(63)
+scalar taintValue extends string;
+
+/** labelValue is the k8s valid value of the label on the nodepool nodes
+ * The good example of the label value is `master`
+ */
+@minLength(1)
+@maxLength(63)
+scalar labelValue extends string;
```
Comment (mjlshen) [inline api/redhatopenshift/HcpCluster/hcpCluster-models.tsp:462]: Will we need a regex restriction here? https://kubernetes.io/docs/reference/kubectl/generated/kubectl_taint/

* The key must begin with a letter or number, and may contain letters, numbers, hyphens, dots, and underscores, up to 253 characters.
* Optionally, the key can begin with a DNS subdomain prefix and a single '/', like example.com/my-app.
* The value is optional. If given, it must begin with a letter or number, and may contain letters, numbers, hyphens, dots, and underscores, up to 63 characters.

File: api/Makefile (hunk 7 above)
Comment (mjlshen) [reply api/Makefile:7]: I'll pull this into #82 instead to simplify things

Comment (mjlshen) [reply api/Makefile:7]: Fixed - we can remove the changes to this Makefile in this PR now :D

File: api/redhatopenshift/HcpCluster/hcpCluster-models.tsp (hunk 8 above)
Comment (petrkotas) [reply api/redhatopenshift/HcpCluster/hcpCluster-models.tsp:462]: We talked about this with @miguelsorianod. We will validate later on the same as is done internally in k8s. Originally I have written the regex, but covering all edge cases made that regex non maintainable.

Comment (mjlshen) [reply api/redhatopenshift/HcpCluster/hcpCluster-models.tsp:462]: Thanks, I'm ok with that - how do we normally keep track of that, a TODO comment or follow up card?

File: internal/api/v20240610preview/generated/constants.go [hunk 9]
```diff
@@ -46,6 +46,27 @@ func PossibleCreatedByTypeValues() []CreatedByType {
...
+	// EffectNoExecute - NoExecute taint effect
+	EffectNoExecute Effect = "NoExecute"
+	// EffectNoSchedule - NoSchedule taint effect
+	EffectNoSchedule Effect = "NoSchedule"
+	// EffectPreferNoSchedule - PreferNoSchedule taint effect
+	EffectPreferNoSchedule Effect = "PreferNoSchedule"
+)
+
+// PossibleEffectValues returns the possible values for the Effect const type.
+func PossibleEffectValues() []Effect {
```
Comment (s-amann) [inline internal/api/v20240610preview/generated/constants.go:62]: does this need to be added to `register.go`?

**Review by s-amann — APPROVED** [review]

File: internal/api/v20240610preview/generated/constants.go (hunk 9 above)
Comment (petrkotas) [reply internal/api/v20240610preview/generated/constants.go:62]: Good catch, I did not saw the register. Fixed.

**Review by s-amann — APPROVED** [review]
